docs/adr/2023-11-07_029-streaming-persistence.md

```diff
@@ -0,0 +1,72 @@
+ - An application that wanted a deeper view into the workings of the hydra protocol would have no access to several of the internal event types 
+
+Additionally, much of the changes that would need to happen to solve any of these would apply equally well to the on-disk persistence the Hydra node currently provides.
+
+-------------------------------------
+
+- Current client API is a stateful WebSocket connection. Client API has its own incrementally persisted state that must be maintained, and includes a monotonically increasing `Seq` ID, but this is based off the events a client observes, not the Hydra host's event log.
+
+- Client must be flexible and ready to handle many different events
+
+- There is no way to simply hand off transactions to the hyrda-node currently, a full connection must be initiated before observed transactions can be applied, and bulky JSON objects can slow down "bulk" operations
+
+- Many applications, like SundaeSwap's Gummiworm Protocol, do not need need the full general API, and benefit from any performance improvement
+
+- The challenge of subscribing to event types is complex to handle, but would be applicable to many parts of Hydra, not just for subscribing to new transactions. It's a good fit for passing stuff off to a message queue (MQTT, Kafka, whatever), probably from a dedicated process (see "Chain Server" below)
+    - Could also just directly use a "compatible" spec like STOMP or MQTT, but that would lock in implementation details
+
+- Previous mock chain using ZeroMQ was [removed][zmq] as part of [#119][#119], due to complexity and feeling of ZeroMQ being unmaintained (primary author no longer contributing, new tag release versions not being published)
+    - This mock chain was used to mock the layer L1 chain, not the L2 ledger
+- Attempt in February 2023 to externalize chainsync server as part of [#230][#230]
+    - Similar to [#119][#119], but the "Chain Server" component in charge of translating Hydra Websocket messages into direct chain, mock chain, or any hypothetical message queue, not necessarily just ZeroMQ
+    - Deemed low priority due to ambiguous use-case at the time. SundaeSwap's Gummiworm Protocol would benefit from the additional control enabled by the Event Server
+
+# Decision
+
+A new abstraction, the EventServer, will be introduced. Each internal hydra event will be sent to the event server, which will be responsible for persisting that event and returning a durable monotonically increasing global event ID. Different implementations of the event server can handle this event differently. Initially, we will implement the following:
+ - MockEventServer, which increments a counter for the ID, and discards the event
+ - FileEventServer, which increments a counter for the ID, and encapsulates the existing file persistence mechanism
+ - SocketEventServer, which increments a counter for the ID, and writes the event to an arbitrary unix socket
```

Comment:
Namely, it could write to a unix socket in a binary format, rather than incurring the overhead of a web socket + tcp + serialization. We can also expose *all* internal events to the socket, which would be useful in ecosystem integrations, whereas the websocket currently exposes only a subset of events (from my understanding, I may be wrong here)

Reply:
The websocket currently contains all information, even a bit more. The shape is different and additional persistence is hacky. So it's not perfect, but we should be able to unify things here.

I don't think there is much of an overhead in using WebSockets over unix domain sockets. Plus there is additional complexity of message framing on domain sockets.

It sounds like what you want is rather a more compact encoding (binary instead of JSON)? I have suggested a way to select the encoding via a content-type in [ADR25](https://hydra.family/head-protocol/adr/25/)

Reply:
Ok; my understanding from talking with Arnaud was that there were some events that are modelled internally but not exposed to the websocket, so we were trying to provide a deeper integration without breaking the websocket API, but yea, in that case we can unify the websocket and unix socket APIs for now